### PR TITLE
Add more NullPointerException throw sites

### PR DESCRIPTION
**Pull request by smowton**
NPE can already be emitted on a field access against null (e.g. `((A)null).field`). This adds NPE throws on array access, e.g. `((int[])null)[0]`, and virtual call, e.g. `((Object)null).equals(null)`, and adds tests to ensure that they can be caught.

### CODE REVIEW

**Comment by romainbrenguier** [comment]
I seems one the test `StringMiscellaneous04` is now failing because it can now throw `null`.
If this is correct behaviour then you can adapt the test description.

File: regression/cbmc-java/exceptions24/test.desc [hunk 1]
```diff
@@ -0,0 +1,12 @@
+CORE
+test.class
+--java-throw-runtime-exceptions
+^EXIT=10$
+^SIGNAL=0$
+VERIFICATION FAILED
```
Comment (romainbrenguier) [inline regression/cbmc-java/exceptions24/test.desc]: We should have some positive tests as well, to show that function that would not throw exception are detected by CBMC as such.

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 2]
```diff
@@ -464,23 +477,17 @@ codet java_bytecode_instrumentt::instrument_expr(
...
-          dereference_expr.op0(),
-          dereference_expr.source_location(),
-          false);
-      return null_dereference_check;
-    }
+    // Check pointer non-null before access:
+    const dereference_exprt dereference_expr=to_dereference_expr(expr);
+    codet null_dereference_check=
+      check_null_dereference(
+        dereference_expr.op0(),
```
Comment (romainbrenguier) [inline src/java_bytecode/java_bytecode_instrument.cpp]: use `.pointer()`

**Comment by smowton** [comment]
CI is failing due to dependency on https://github.com/diffblue/cbmc/pull/1214

**Pull request merged**